comp/dogstatsd/listeners/connections_tracker.go

```diff
@@ -0,0 +1,136 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// Package listeners implements the StatsdListener interfaces.
+package listeners
+
+import (
+	"net"
+	"time"
+
+	"go.uber.org/atomic"
+
+	"github.com/DataDog/datadog-agent/pkg/trace/log"
+	logger "github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// ConnectionTracker tracks connections and closes them gracefully.
+type ConnectionTracker struct {
+	connections       map[net.Conn]struct{}
+	connToTrack       chan net.Conn
+	connToClose       chan net.Conn
+	stopChan          chan struct{}
+	stoppedChan       chan struct{}
+	closeDelay        time.Duration
+	activeConnections *atomic.Int32
+	name              string
+}
+
+// NewConnectionTracker creates a new ConnectionTracker.
+// closeDelay is the time to wait before closing a connection. First it will be shutdown for write, which
+// will notify the client that we are disconnecting, then it will be closed. This gives some time to
+// consume the remaining packets.
+func NewConnectionTracker(name string, closeDelay time.Duration) *ConnectionTracker {
+	return &ConnectionTracker{
+		connections:       make(map[net.Conn]struct{}),
+		connToTrack:       make(chan net.Conn),
+		connToClose:       make(chan net.Conn),
+		stopChan:          make(chan struct{}),
+		stoppedChan:       make(chan struct{}),
+		closeDelay:        closeDelay,
+		activeConnections: atomic.NewInt32(0),
+		name:              name,
+	}
+}
+
+// Start starts the connection tracker.
+func (t *ConnectionTracker) Start() {
+	go t.HandleConnections()
+}
+
+// Track tracks a connection.
+func (t *ConnectionTracker) Track(conn net.Conn) {
+	t.connToTrack <- conn
+}
+
+// Close closes a connection.
+func (t *ConnectionTracker) Close(conn net.Conn) {
+	t.connToClose <- conn
+}
+
+// HandleConnections handles connections.
+func (t *ConnectionTracker) HandleConnections() {
+	requestStop := false
+	for stop := false; !stop; {
+		select {
+		case conn := <-t.connToTrack:
+			log.Debugf("dogstatsd-%s: tracking new connection %s", t.name, conn.RemoteAddr().String())
+
+			t.connections[conn] = struct{}{}
+			t.activeConnections.Inc()
+		case conn := <-t.connToClose:
+			err := conn.Close()
+			log.Infof("dogstatsd-%s: failed to close connection: %v", t.name, err)
+
+			delete(t.connections, conn)
+			t.activeConnections.Dec()
+			if requestStop && len(t.connections) == 0 {
```

Comment:
I think this is racy and will deadlock if, assuming connections is not empty, a new connection is accepted shortly before `Stop()`, and `connToTrack` branch adds a new connection between `stopChan` and `connToClose` branches run.

Reply:
That's true, if the contract is that we close the listening socket before then it should be fine no ?
Any other suggestion ?